### PR TITLE
Implement Unix.truncate and Unix.ftruncate on Windows

### PR DESCRIPTION
**Pull request by nojb**
This is a follow-up to #1939. That PR seems to have been abandoned by the original author; in order not to lose the work contained there, I rebased it on top of `trunk`, cleaned it up, added an implementation for `Unix.ftruncate` and added some tests, which I am now submitting for review.

@dra27 would you mind reviewing this?

Will clean up the commit history if this is accepted for inclusion.

### CODE REVIEW

**Comment by fccm** [comment]
Thanks Nicolás for having continued the work (I don't think I would have been able to do what you've done). I hope it will find its way for inclusion.

Would it be impolite if I make a call for review on the ocaml mailing list?

**Comment by nojb** [comment]
> Would it be impolite if I make a call for review on the ocaml mailing list?

I requested a review from @dra27 which should be enough. Thanks anyway!

File: otherlibs/win32unix/truncate.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           */
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: This needs updating!

File: otherlibs/win32unix/truncate.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+/*                                                                        */
+/*   Copyright 2018 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c:9]: Similarly.

File: otherlibs/win32unix/truncate.c [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/signals.h>
+#include <caml/io.h>
+#include <caml/osdeps.h>
+#include "unixsupport.h"
+#include <windows.h>
+
+static int win_truncate_handle(HANDLE fh, long long len)
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]:
```suggestion
static int win_truncate_handle(HANDLE fh, __int64 len)
```
(I'm not certain about the portability of `long long` to the older compilers, so it seems simpler to use the Windows type which encapsulates it - there's `ARCH_INT64_TYPE` defined in `config.h` or just the fact that other files in win32unix.c use `__int64`)

File: otherlibs/win32unix/truncate.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  LARGE_INTEGER fp;
+  fp.QuadPart = len;
+  if (SetFilePointerEx(fh, fp, NULL, FILE_BEGIN) == 0 ||
+      SetEndOfFile(fh) == 0) {
+    return -1;
+  }
+  return 0;
+}
+
+static int win_ftruncate(HANDLE fh, long long len)
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]:
```suggestion
static int win_ftruncate(HANDLE fh, __int64 len)
```

File: otherlibs/win32unix/truncate.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  if (DuplicateHandle(currproc, fh, currproc, &dupfh, 0, FALSE,
+                      DUPLICATE_SAME_ACCESS) == 0) {
+     return -1;
+  }
+  ret = win_truncate_handle(dupfh, len);
+  CloseHandle(dupfh);
+  return ret;
+}
+
+static int win_truncate(WCHAR * path, long long len)
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]:
```suggestion
static int win_truncate(WCHAR * path, __int64 len)
```

File: otherlibs/win32unix/truncate.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    return -1;
+  }
+  return 0;
+}
+
+static int win_ftruncate(HANDLE fh, long long len)
+{
+  HANDLE dupfh, currproc;
+  int ret;
+  if (fh == INVALID_HANDLE_VALUE) {
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: FWIW, why bother with this - `DuplicateHandle` will fail

File: otherlibs/win32unix/truncate.c [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  caml_leave_blocking_section();
+  caml_stat_free(p);
+  if (ret == -1)
+    uerror("truncate", path);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_ftruncate(value fd, value len)
+{
+  CAMLparam1(len);
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]:
```suggestion
  CAMLparam2(fd, len);
```
Equally, these functions don't allocate or do anything with the parameters after the blocking section, so you could just remove the `CAMLparam` completely and just have `return Val_unit;` at the end.

File: otherlibs/win32unix/truncate.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  int ret;
+  caml_enter_blocking_section();
+  ret = win_ftruncate(Handle_val(fd), Long_val(len));
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: `Handle_val` cannot be used between `caml_enter_blocking_section`..`caml_leave_blocking_section`

File: otherlibs/win32unix/truncate.c [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_ftruncate_64(value fd, value vlen)
+{
+  CAMLparam1(vlen);
+  int ret;
+  file_offset len = File_offset_val(vlen);
+  caml_enter_blocking_section();
+  ret = win_ftruncate(Handle_val(fd), len);
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: As in `unix_ftruncate`, the `Handle_val` needs hoisting

File: otherlibs/win32unix/truncate.c [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  ret = win_ftruncate(Handle_val(fd), Long_val(len));
+  caml_leave_blocking_section();
+  if (ret == -1)
+    uerror("ftruncate", Nothing);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_ftruncate_64(value fd, value vlen)
+{
+  CAMLparam1(vlen);
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: Either `CAMLparam2` or eliminate completely.

File: otherlibs/win32unix/truncate.c [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+  int ret;
+  file_offset len = File_offset_val(vlen);
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: Why not use `Int64_val` and `__int64`?

File: otherlibs/win32unix/truncate.c [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    uerror("truncate", path);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_truncate_64(value path, value vlen)
+{
+  CAMLparam2(path, vlen);
+  WCHAR * p;
+  int ret;
+  file_offset len = File_offset_val(vlen);
```
Comment (dra27) [inline otherlibs/win32unix/truncate.c]: Why not use `Int64_val` and `__int64`?

**Review by dra27 — CHANGES_REQUESTED** [review]
There's a GC issue in the `ftruncate` implementation and other than that just a few C nits.

**Comment by dra27** [comment]
@shindere - the tests in this PR fail on Unix because the files `truncate.txt` and `ftruncate.txt` are symlinked into the test directory instead of being copied and they are altered by the test. Is there an easy way of telling ocamltest to copy them always (not just on Windows)?

**Comment by shindere** [comment]
David Allsopp (2019/05/02 15:16 +0000):
> @shindere - the tests in this PR fail on Unix because the files
> `truncate.txt` and `ftruncate.txt` are symlinked into the test
> directory instead of being copied and they are altered by the test. Is
> there an easy way of telling ocamltest to copy them always (not just
> on Windows)?

Well, not in a declarative way.
The solution would be to not declare them in the files variable, but
rather to use a script to copy them from the source to the test
directory.

**Comment by dra27** [comment]
@shindere - OK, the problem is that then the test script becomes much more verbose. I think for these tests what will be easier is for the test script to write the files instead. I wonder if it would be useful to have a declarative way for some of the other tests - I'll open an issue, as obviously that's something to track, not something which requires urgent fixing.

**Comment by shindere** [comment]
David Allsopp (2019/05/03 01:49 -0700):
> @shindere - OK, the problem is that then the test script becomes much
> more verbose.

Really? The way I imagined it, it's just a matter of writing a shell
script that does all the copies and invoke it through a "script" test.
Is that really too verbose, or am I missing something?

> I think for these tests what will be easier is for the test script to
> write the files instead.

It could definitely work that way, indeed. And the writing could then
even be done in OCaml.

**Comment by nojb** [comment]
@fccm This contribution is not small enough to pass under the "no copyright" policy, so you will need to sign a CLA and send it to @xavierleroy. This is all explained in [CONTRIBUTING.md#Contributor License Agreement](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#contributor-license-agreement).

**Comment by dra27** [comment]
> David Allsopp (2019/05/03 01:49 -0700):
> >  @shindere - OK, the problem is that then the test script becomes much
> >  more verbose.

> Really? The way I imagined it, it's just a matter of writing a shell
> script that does all the copies and invoke it through a "script" test.
> Is that really too verbose, or am I missing something?

Isn't it then necessary to explicitly run the `ocamlc` and `ocamlopt` actions, though? Otherwise the test just becomes the `cp` script?

File: otherlibs/win32unix/truncate.c (hunk 1 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done.

File: otherlibs/win32unix/truncate.c (hunk 2 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c:9]: Done.

File: otherlibs/win32unix/truncate.c (hunk 3 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done, switched to `__int64`.

File: otherlibs/win32unix/truncate.c (hunk 4 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done.

File: otherlibs/win32unix/truncate.c (hunk 5 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done.

File: otherlibs/win32unix/truncate.c (hunk 7 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done, got rid of the CAML macros altogether.

File: otherlibs/win32unix/truncate.c (hunk 8 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Good catch, fixed.

File: otherlibs/win32unix/truncate.c (hunk 9 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Fixed.

File: otherlibs/win32unix/truncate.c (hunk 10 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Done, got rid of the macros.

File: otherlibs/win32unix/truncate.c (hunk 11 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Indeed, fixed.

File: otherlibs/win32unix/truncate.c (hunk 12 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Indeed, fixed here as well.

File: otherlibs/win32unix/truncate.c (hunk 6 above)
Comment (nojb) [reply otherlibs/win32unix/truncate.c]: Indeed, I removed the check.

**Comment by shindere** [comment]
David Allsopp (2019/05/03 09:05 +0000):
> Isn't it then necessary to explicitly run the `ocamlc` and `ocamlopt`
> actions, though? Otherwise the test just becomes the `cp` script?

If you are fine with the `native` and `bytecode` tests, you could run
them -- they are the ones that are run by default

**Comment by dra27** [comment]
I just pushed a revision to the test (which you're free to revise further/replace, of course!) - it writes the file instead of copying it but also eliminates the newline which means that it will/should work on a CRLF checkout too.

**Comment by nojb** [comment]
> I just pushed a revision to the test (which you're free to revise further/replace, of course!) - it writes the file instead of copying it but also eliminates the newline which means that it will/should work on a CRLF checkout too.

Thanks, was about to do the same, but you beat me to it! 😄

**Comment by nojb** [comment]
I think all issues have been addressed (modulo CLA for @fccm).

**Comment by shindere** [comment]
David Allsopp (2019/05/03 02:22 -0700):
> I just pushed a revision to the test (which you're free to revise
> further/replace, of course!) - it writes the file instead of copying
> it but also eliminates the newline which means that it will/should
> work on a CRLF checkout too.

Didn't you make ocamltest's file comparison code robust to CRLF
checkouts anyway?

The test looks good to me! Thanks!

**Comment by fccm** [comment]
@nojb I just sent the CLA-individual.pdf filled and signed to Xavier Leroy by email.
Thanks all again for the work done :)

**Comment by nojb** [comment]
> @nojb I just sent the CLA-individual.pdf filled and signed to Xavier Leroy by email.
> Thanks all again for the work done :)

Thanks! Once @xavierleroy confirms receipt of the CLA and @dra27 approves the PR, this can be merged.

**Comment by dra27** [comment]
@shindere:
> >  I just pushed a revision to the test (which you're free to revise
>  further/replace, of course!) - it writes the file instead of copying
>  it but also eliminates the newline which means that it will/should
> work on a CRLF checkout too.

> Didn't you make ocamltest's file comparison code robust to CRLF
> checkouts anyway?

This was an issue with the test itself, not ocamltest - prior to the change, the test file contained `Hello, OCaml!\n` and then read the file length, so on a CRLF checkout the ending would have been changed by Git to be `\r\n`.

> The test looks good to me! Thanks!

Thanks!

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
This all LGTM! Up to you @nojb if you want to produce a clean history or just squash this on to trunk.

**Comment by nojb** [comment]
I squashed-merged. Thanks!

**Comment by xavierleroy** [comment]
CLA received in good order.

**Pull request merged**